soc/nordic/common/dmm.c

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+int dmm_dma_buffer_out_prepare(const struct device *dev, void const * user_buffer, void ** buffer_out)
+{
+    // Get memory region that specified device can perform DMA transfers from, using devicetree
+
+    // Check if provided user buffer is already in correct memory region
+    // If yes, assign buffer_out to user_buffer
+    // If no:
+    // - dynamically allocate buffer in correct memory region, using for example mem_attr_heap_alloc()
+    // - copy user buffer contents into allocated buffer
+
+    // Return error if dynamic allocation fails
+
+    // Check if device memory region is cacheable
+    // If yes, writeback all cache lines associated with output buffer (either user or allocated)
+    // If no, no action is needed
+}
+
+int dmm_dma_buffer_out_release(void const * buffer_out)
+{
+    // Get memory region that specified device can perform DMA transfers from, using devicetree
```

Comment:
Is this step needed here?

Reply:
So the function should get the memory region controlled by the dynamic allocator, which is not necessarily the same as the one from which DMA can perform a transfer. If the API user passes a static DMAable buffer it must not be freed in this function.

I asked if this action was needed because I assumed the check in the next line implicitly gets the memory range supervised by the allocator to compare the buffer address with the dynamic allocator range.

Reply:
so we are missing `dev` parameter - once this is available we can perform the action you described

Reply:
Yes, I'm not sure it's required in all cases, but I can imagine scenarios in which it might.
Also, adding a `dev` parameter is good for function parameters symmetry.